File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(),nums.end());
        vector<vector<int>>v;
        int i,lo,hi,n=nums.size();
        for(i=0;i<n;i++){
            lo=i+1;hi=n-1;
            while(lo<hi){
                int sum=nums[i]+nums[lo]+nums[hi];
                if(sum==0){
                    v.push_back({nums[i],nums[lo],nums[hi]});
                    while(lo<hi and nums[lo]==nums[lo+1])lo++;
                    while(lo<hi and nums[hi]==nums[hi-1])hi--;
        
                    lo++;hi--;}
                else if(sum>0){
                    hi--;
                }
                else lo++;
            }
            while(i+1<n and nums[i+1]==nums[i]) i++;
        }
        return v;
    }
};

File: 0015-3sum/0015-3sum.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var threeSum = function (nums) {
    const n = nums.length;
    nums.sort((a, b) => a - b);
    const res = [];

    for (let i = 0; i < n - 1; i++) {
        if (i > 0 && nums[i] === nums[i - 1]) continue;

        let l = i + 1, r = n - 1;

        while (l < r) {
            const target = nums[i] + nums[l] + nums[r];
            if (target === 0) {
                res.push([nums[i], nums[l], nums[r]]);
                while (l < r && nums[l] === nums[l + 1]) l++;
                while (l < r && nums[r] === nums[r - 1]) r--;
                l++, r--;
            }
            else if (target > 0) r--;
            else l++;
        }

    }

    return res;
};